File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        String[] input = br.readLine().split(" ");
        int[] numbers = new int[N + 1];
        List<Integer>[] check = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            check[i] = new ArrayList<>();
            numbers[i] = Integer.parseInt(input[i - 1]);
        }
        int answer = 0;
        int index = 0;
        int[] dp = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            dp[i] = 1;
            int max = 0;
            for (int j = 1; j < i; j++) {
                if (numbers[i] > numbers[j]) {
                    if (dp[j] > max) {
                        max = dp[j];
                        check[i].clear();
                        check[i].addAll(check[j]);
                    }
                }
            }
            check[i].add(numbers[i]);
            dp[i] = max + 1;
            if (dp[i] > answer) {
                answer = dp[i];
                index = i;
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int num : check[index]) {
            sb.append(num).append(" ");
        }
        System.out.println(answer);
        System.out.println(sb.toString());
    }
}

File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.py
n = int(input())
nums = list(map(int, input().split()))

# 가장 처음의 수는 길이를 1로 갖는다.
# 이전 수와 비교해서 가장 긴 길이에 이어서 길이를 추가한다.
# 해당 문제는 가장 긴 수열도 포함해서 제공해야한다.

# 길이랑 해당 원소 넣기
dp = [0] * len(nums)
check = [[]] * len(nums)

for i in range(0, len(nums)):
    dp[i] = 1
    num = nums[i]
    check[i] = [num]
    for j in range(0, i):
        if num > nums[j]:
            if dp[j] + 1 > dp[i]:
                dp[i] = dp[j] + 1
                update = check[j][:]
                update.append(num)
                check[i] = update

max_len = 0
max_index = 0

for i in range(len(dp)):
    if dp[i] > max_len:
        max_len = dp[i]
        max_index = i

print(dp[max_index])
print(" ".join(map(str,check[max_index])))
